// nacos-config/dist/interface.js
"use strict";
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// import { EventEmitter } from 'events';
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientOptionKeys = void 0;
var ClientOptionKeys;
(function (ClientOptionKeys) {
    ClientOptionKeys["ENDPOINT"] = "endpoint";
    ClientOptionKeys["SERVER_PORT"] = "serverPort";
    ClientOptionKeys["NAMESPACE"] = "namespace";
    ClientOptionKeys["ACCESSKEY"] = "accessKey";
    ClientOptionKeys["SECRETKEY"] = "secretKey";
    ClientOptionKeys["HTTPCLIENT"] = "httpclient";
    ClientOptionKeys["APPNAME"] = "appName";
    ClientOptionKeys["SSL"] = "ssl";
    ClientOptionKeys["SNAPSHOT"] = "snapshot";
    ClientOptionKeys["CACHE_DIR"] = "cacheDir";
    ClientOptionKeys["NAMESERVERADDR"] = "nameServerAddr";
    ClientOptionKeys["SERVERADDR"] = "serverAddr";
    ClientOptionKeys["UNIT"] = "unit";
    ClientOptionKeys["REFRESH_INTERVAL"] = "refreshInterval";
    ClientOptionKeys["CONTEXTPATH"] = "contextPath";
    ClientOptionKeys["CLUSTER_NAME"] = "clusterName";
    ClientOptionKeys["REQUEST_TIMEOUT"] = "requestTimeout";
    ClientOptionKeys["HTTP_AGENT"] = "httpAgent";
    ClientOptionKeys["SERVER_MGR"] = "serverMgr";
    ClientOptionKeys["DEFAULT_ENCODING"] = "defaultEncoding";
    ClientOptionKeys["IDENTITY_KEY"] = "identityKey";
    ClientOptionKeys["IDENTITY_VALUE"] = "identityValue";
    ClientOptionKeys["DECODE_RES"] = "decodeRes";
    ClientOptionKeys["ENDPOINT_QUERY_PARAMS"] = "endpointQueryParams";
})(ClientOptionKeys = exports.ClientOptionKeys || (exports.ClientOptionKeys = {}));
